core/libraries/plugin_api/EE_Register_Capabilities.lib.php

```diff
@@ -114,25 +108,26 @@ public static function register($cap_reference = null, $setup_args = array())
      * @param array $incoming_caps The original caps map.
      * @return array merged in new caps.
      */
-    public static function register_capabilities($incoming_caps)
+    public static function register_capabilities(array $incoming_caps): array
     {
+        $caps = [];
         foreach (self::$_registry as $cap_reference => $caps_and_cap_map) {
-            $incoming_caps = array_merge_recursive($incoming_caps, $caps_and_cap_map['caps']);
+            $caps[] = $caps_and_cap_map['caps'];
         }
-        return $incoming_caps;
+        return array_merge_recursive($incoming_caps, $caps);
```

Comment:
i replaced a bunch of uses of `array_merge()` within `foreach` loops because that is considered incredibly inefficient and memory intensive because of all of the additional "hidden" arrays that get created behind the scenes.